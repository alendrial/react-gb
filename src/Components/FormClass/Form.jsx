import React, { Component, Fragment } from "react"
import { Input } from "./Input"
import { Button } from "./Button"
import { Message } from "./Message"

export class Form extends Component {
  state = {
    name: "Click here!",
    value: "",
    messages: [],
    visible: true,
  }

  handleClick = () => {
    this.setState({ messages: [...this.state.messages, this.state.value] })
    this.setState({ value: "" })
  }

  handleChange = (event) => {
    this.setState({ value: event.target.value })
  }

  render() {
    return (
      <Fragment>
        <Input change={this.handleChange} value={this.state.value} />
        <br />
        <Button
          className="add-btn"
          name={this.state.name}
          click={this.handleClick}
          type="submit"
        />
        <button
          className="hide-btn"
          onClick={() => this.setState({ visible: !this.state.visible })}
        >
          {this.state.visible ? "Hide messages" : "Show messages"}
        </button>
        <button
          className="rm-btn"
          onClick={() => this.setState({ messages: [] })}
        >
          Clear messages
        </button>
        {this.state.visible && <Message post={this.state.messages} />}
      </Fragment>
    )
  }
}
