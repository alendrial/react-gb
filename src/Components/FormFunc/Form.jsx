import React, {
  useState,
  Fragment,
  useCallback,
  useRef,
  useEffect,
} from "react"
import { Input } from "./Input"
import { Button } from "./Button"
import { InputAuthor } from "./InputAuthor"
import { Chat } from "./Chat"

export const Form = () => {
  const [name, setName] = useState("Send message!")

  const [defaultMessage] = useState("Your message has been received!")
  const [defaultAuthor] = useState("Call me mr. Robot!")

  const [message, setMessage] = useState("")
  const [author, setAuthor] = useState("")
  const [messagesList, setMessagesList] = useState([])
  const [time, setTime] = useState("")

  const [visible, setVisible] = useState(true)

  const createCurrentTime = () => {
    let time = new Date()
    return `${time.getHours()} : ${
      (time.getMinutes() < 10 ? "0" : "") + time.getMinutes()
    }`
  }

  const handleClick = () => {
    setMessagesList([
      ...messagesList,
      { message, author, time, robotMessage: false },
    ])
    setMessage("")
    setAuthor("")
    setTime("")
  }

  const changeM = useCallback((event) => {
    setMessage(event.target.value)
  }, [])

  const changeAu = useCallback((event) => {
    setAuthor(event.target.value)
    setTime(createCurrentTime())
  }, [])

  useEffect(() => {
    if (
      messagesList.length &&
      !messagesList[messagesList.length - 1].robotMessage
    ) {
      alert("Wait for your response, it will take a few seconds")
      const timeout = setTimeout(() => {
        setMessagesList([
          ...messagesList,
          {
            message: defaultMessage,
            author: defaultAuthor,
            time: `${createCurrentTime()}`,
            robotMessage: true,
          },
        ])
      }, 1000)
      return () => {
        clearTimeout(timeout)
      }
    }
  }, [messagesList])

  return (
    <Fragment>
      <div className="FormFlexBox">
        <Input change={changeM} message={message} />

        <InputAuthor change={changeAu} author={author} />
      </div>

      <div className="ButtonFlexBox">
        <Button
          className="add-btn"
          name={name}
          click={handleClick}
          type="submit"
          disabled={message && author ? false : true}
        />

        <button className="hide-btn" onClick={() => setVisible(!visible)}>
          {visible ? "hide" : "show"}
        </button>

        <button className="rm-btn" onClick={() => setMessagesList([])}>
          Clear messages
        </button>
      </div>

      {visible && <Chat post={messagesList} />}
    </Fragment>
  )
}
