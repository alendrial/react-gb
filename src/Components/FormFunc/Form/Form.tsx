import React, {
  useState,
  Fragment,
  useCallback,
  useRef,
  useEffect,
  FC,
} from 'react';
import { Input } from './../InputMessage/Input';
import { Button } from './../Button/Button';
import { InputAuthor } from './../InputAuthor/InputAuthor';
import { Chat } from './../Chat/Chat';
import { Message } from './../Chat/Message/Message'

export interface FormProps {
  defaultMessage: string;
  defaultAuthor: string;
  visible: boolean;
  robotMessage: boolean;
}


export const Form: FC<FormProps> = () => {
  const [name, setName] = useState<string>('Send message!');

  const [defaultMessage] = useState<string>('Your message has been received!');
  const [defaultAuthor] = useState<string>('Call me mr. Robot!');

  const [message, setMessage] = useState<string>('');
  const [author, setAuthor] = useState<string>('');
  const [messagesList, setMessagesList] = useState<Message[]>([]);
  const [time, setTime] = useState<string>('');

  const [visible, setVisible] = useState(true);

  const createCurrentTime = () => {
    let time = new Date();
    return `${time.getHours()} : ${
      (time.getMinutes() < 10 ? '0' : '') + time.getMinutes()
    }`;
  };

  const handleClick = () => {
    setMessagesList([
      ...messagesList,
      { message, author, time, robotMessage: false },
    ]);
    setMessage('');
    setAuthor('');
    setTime('');
  };

  const changeM = useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
    setMessage(event.target.value);
  }, []);

  const changeAu = useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
    setAuthor(event.target.value);
    setTime(createCurrentTime());
  }, []);

  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    inputRef.current!.focus()
}, [messagesList])

  useEffect(() => {
    if (
      messagesList.length &&
      !messagesList[messagesList.length - 1].robotMessage
    ) {
      alert('Wait for your response, it will take a few seconds');
      const timeout = setTimeout(() => {
        setMessagesList([
          ...messagesList,
          {
            message: defaultMessage,
            author: defaultAuthor,
            time: `${createCurrentTime()}`,
            robotMessage: true,
          },
        ]);
      }, 1000);
      return () => {
        clearTimeout(timeout);
      };
    }
  }, [messagesList]);

  return (
    <Fragment>
      <div className="formFlexBox">
        <Input change={changeM} message={message} ref={inputRef} />

        <InputAuthor change={changeAu} author={author} />
      </div>

      <div className="buttonFlexBox">
        <Button
          className="add-btn"
          name={name}
          click={handleClick}
          type="submit"
          disabled={message && author ? false : true}
        />

        <button className="hide-btn" onClick={() => setVisible(!visible)}>
          {visible ? 'hide' : 'show'}
        </button>

        <button className="rm-btn" onClick={() => setMessagesList([])}>
          Clear messages
        </button>
      </div>

      {visible && <Chat post={messagesList} />}
    </Fragment>
  );
};
