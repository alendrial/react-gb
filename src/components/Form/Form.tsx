import React, { useState, FC, memo } from 'react';
import { Input } from '../InputMessage/Input';
import { Button } from '../Button/Button';

interface FormProps {
  addMessage: (a: string) => void;
}

export const Form: FC<FormProps> = memo(({ addMessage }) => {
  const [value, setValue] = useState('');

  const handleSubmitForm = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    addMessage(value);
    setValue('');
  };

  return (
    <form onSubmit={handleSubmitForm}>
      <Input
        placeholder="Enter something..."
        value={value}
        onChange={(event) => setValue(event.target.value)}
      />
      <Button disabled={!value} />
    </form>
  );
});
