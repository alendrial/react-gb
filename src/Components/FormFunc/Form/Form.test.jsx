import React from 'react';
import { Form } from '../Form/Form';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { toBeDisabled } from '@testing-library/jest-dom/dist/matchers';
import userEvent from '@testing-library/user-event';
import { waitFor } from '@storybook/testing-library';

describe('Form testing render with snapshot', (props) => {
  it('should render the form with snapshot', () => {
    const { asFragment } = render(<Form {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
