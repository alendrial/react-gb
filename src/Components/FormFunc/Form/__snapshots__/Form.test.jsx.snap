// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form testing render with snapshot should render the form with snapshot 1`] = `
<DocumentFragment>
  <div
    class="FormFlexBox"
  >
    <input
      class="InputText"
      placeholder="Enter something..."
      style="margin-right: 20px;"
      type="text"
      value=""
    />
    <input
      class="InputAuthor"
      placeholder="Username"
      type="text"
      value=""
    />
  </div>
  <div
    class="buttonFlexBox"
  >
    <button
      disabled=""
      style="background-color: rgb(76, 196, 156);"
    >
      Send message!
    </button>
    <button
      class="hide-btn"
    >
      hide
    </button>
    <button
      class="rm-btn"
    >
      Clear messages
    </button>
  </div>
  <div
    class="chatBox"
  >
    <ul
      class="msg-block"
      style="list-style: none;"
    />
  </div>
</DocumentFragment>
`;
